### ISSUE TITLE
Am I training properly?

### ISSUE THREAD

**Issue by resurgo97** [openspeech-team/openspeech#130]
I am not familiar with ASR tasks, so I'd be so glad if anyone answer my question:

I am training ContextNet which is basically rnn-t type model as in the original paper.
Because I need only encoder part of the model, I am using 'contextnet' instead of 'contextnet_transducer',
which helps train faster & reduce memory usage.
Since I only have encoder, I also use as criterion 'ctc' instead of 'rnnt'.

I am not sure if this configuration is valid for proper training.

![image](https://user-images.githubusercontent.com/37990297/144692582-3764e666-8d56-472d-b547-f67d5c26e8ff.png)

Now at epoch 8, valid_wer=0.996 and valid_cer=7.140, but I doubt they are in expected range though.

---

My training script is as below:

python ./openspeech_cli/hydra_train.py 
dataset=librispeech dataset.dataset_path=/home/ubuntu/TEST/libri 
dataset.dataset_download=False dataset.manifest_file_path=/home/ubuntu/TEST/libri/LibriSpeech/libri_subword_manifest.txt  
tokenizer=libri_subword 
model=contextnet 
audio=fbank 
lr_scheduler=warmup_reduce_lr_on_plateau 
trainer=gpu 
criterion=ctc 
tokenizer.vocab_path=/home/ubuntu/TEST/libri/LibriSpeech/ 
trainer.sampler=random 
lr_scheduler.peak_lr=0.0025 
audio.frame_length=25.0 
trainer.batch_size=128

**Comment by upskyy** [comment]
I think the training is not done correctly, can I see the loss graph?

**Comment by yunigma** [comment]
Hello
I guess I have the same issue with
```model:
  model_name: contextnet_lstm
  model_size: medium
  input_dim: 80
  num_encoder_layers: 5
  num_decoder_layers: 2
  kernel_size: 5
  num_channels: 256
  encoder_dim: 640
  num_attention_heads: 8
  attention_dropout_p: 0.1
  decoder_dropout_p: 0.1
  max_length: 128
  teacher_forcing_ratio: 1.0
  rnn_type: lstm
  decoder_attn_mechanism: loc
  optimizer: adam```
  
  WER does not improve after a few epoches.
Iuliia

**Comment by yunigma** [comment]
@resurgo97 hello. Did you manage to fix this issue finally? Thanks. Iuliia

**Comment by sooftware** [comment]
Can you show the log? cc. @upskyy

**Comment by yunigma** [comment]
Hi @sooftware . Attach here. Thank you!
[logs_20220201_2.log](https://github.com/openspeech-team/openspeech/files/7984853/logs_20220201_2.log)

**Comment by sooftware** [comment]
Obviously something weird. @upskyy look at this. Loss too large.  
I think we need to check if lr is being adjusted.

**Comment by sooftware** [comment]
@yunigma Can you attach the command that you used?

**Comment by yunigma** [comment]
Thank you!!
```
python ./openspeech_cli/hydra_train.py dataset=librispeech dataset.dataset_download=False dataset.dataset_path="../../../../database/LibriSpeech/" dataset.manifest_file_path="../../../openspeech/datasets/librispeech/libri_subword_manifest.txt" tokenizer=libri_subword model=contextnet_lstm audio=fbank lr_scheduler=warmup_reduce_lr_on_plateau trainer=gpu criterion=cross_entropy
```

**Comment by sooftware** [comment]
I'll test it. I need to download new data, so please wait a little bit.

**Comment by upskyy** [comment]
I'm really sorry for the late reply.  
I trained the contextnet model with ctc, and it was confirmed that the training worked well.   

<img width="1349" alt="스크린샷 2022-02-27 오후 11 33 00" src="https://user-images.githubusercontent.com/54731898/155886636-bc2691a9-7d2d-4f08-93e4-725bce97c262.png">

<img width="1355" alt="스크린샷 2022-02-27 오후 11 33 17" src="https://user-images.githubusercontent.com/54731898/155886654-33527aa7-5b60-48bf-be82-8f6ecfbab74d.png">
  
```python
python ./openspeech_cli/hydra_train.py \
    dataset=ksponspeech \
    tokenizer=kspon_character \
    model=contextnet \
    audio=fbank \
    lr_scheduler=warmup_reduce_lr_on_plateau \
    trainer=gpu \
    criterion=ctc
```

**Comment by yunigma** [comment]
@upskyy thank you very much for testing! Do you think that the training can improve slower with the librispeech dataset? Or there is some error in the training itself?

**Comment by upskyy** [comment]
@yunigma  I think it's a subword related issue rather than librispeech dataset. I've confirmed that I'm learning with kspon_character, so How about trying it out with libri_character?

**Comment by yunigma** [comment]
Hi @upskyy ! I have finally tried to reproduce the same setup but with librispeech.
```
```
After two days of training (100 epochs) I got these results:
<img width="1004" alt="Screenshot 2022-04-04 at 17 20 10" src="https://user-images.githubusercontent.com/27154358/161696947-814855d5-e5e6-4e5f-a23a-10b348f62c86.png">

I do not why WER went up at some point. Also I see that the global step in my case is different from yours.

**Comment by sooftware** [comment]
@upskyy I think there is an error in the WER calculation process.

**Comment by yunigma** [comment]
Probably, otherwise CER would also go down... But CER was not improving much either. Do you know how long @upskyy was training and with which parameters?

**Comment by yunigma** [comment]
[log_ctc_char.txt](https://github.com/openspeech-team/openspeech/files/8415790/log_ctc_char.txt)

**Comment by upskyy** [comment]
@yunigma  I think I trained for about 36 hours. Detailed parameters are written in the log. 
I wonder what might have made the difference. 😢 
I'll test it out when I have time.

[hydra_train.log](https://github.com/openspeech-team/openspeech/files/8439741/hydra_train.log)

**Comment by Luongtran99** [comment]
Hi @all, I got issues when training model with openspeech. I did not see sp.model file in Librispeech folder. 
Could you help me?

**Comment by ccyang123** [comment]
You may download the related librispeech files below. (from README.md)

|LibriSpeech|character|[[Link]](https://drive.google.com/file/d/1ZTjkTN8hCwVqiygq8LSIyv_V3D1P3njH/view?usp=sharing)|[[Link]](https://drive.google.com/file/d/1qsSgb3z_qCXYglB7JlPjG3JjLqOWCCIv/view?usp=sharing)|-|
|LibriSpeech|subword|[[Link]](https://drive.google.com/file/d/1IY4Qm1uxbKOcdagNyji5zoXDaSV2KevR/view?usp=sharing)|[[Link]](https://drive.google.com/file/d/16551boLSMTUAeidyDQ8bOSbJoQslDaRj/view?usp=sharing)|[[Link]](https://drive.google.com/file/d/1nqfGJ2uBQJEvjnY-GJ7GgZNZxH_AonoY/view?usp=sharing)|

**Comment by ccyang123** [comment]
@yunigma @upskyy   Did you solve your problem? 
I met the same problem when training the squeezeformer netowrk with LibriSpeech. 
I used "libri_character" as tokenizer and the "libri_char_manifest.txt" as manifest_file_path. 
The CER is going down however, WER is not going down. 

Thank you!

**Comment by ccyang123** [comment]
Both CER and WER are going down after 3 epochs training. 
![image](https://user-images.githubusercontent.com/41611304/210311103-f4ae3a39-1ea7-4dc3-aae8-55f905b07295.png)
